### ISSUE TITLE
Discuss input description

### ISSUE THREAD

**Issue by bpross-52n** [opengeospatial/ogcapi-processes#122]
Since the OGC Process API initially mapped XML bindings to JSON, there might have a lot of XML specificities that would benefit to be better suited to REST/JSON, not only the cardinality of inputs.

I think that _dataDescriptionType_ , _complexInputType_  and _literalInputType_ definitions should be reworked in the future, in particular to define those types using JSON types and constraints.

During the Hackaton 2019, we tried to find how to map the whole process description to JSON schema (https://github.com/spacebel/testbed14/wiki/OGC-Hackaton-2019). I think we should partially apply those mapping for defining the inputs/outputs types.

The following JSON Schema properties shall be used to map the minOccurs and maxOccurs definitions:

- 'Required' attribute in case the property is mandatory (minOccurs >= 1)
- 'Array' type in case the property includes multiple values (minOccurs > 1)
- For arrays, 'minItems' and 'maxItems' for providing any specific cardinality
- 'maxItems' shall be null if the number of items is unbounded

```
input1:
  type: array
  minItems: 1
  maxItems: 10
  items:
    $ref: 'myItemType'
```

To be discussed...

**Comment by bpross-52n** [comment]
See https://github.com/opengeospatial/ogcapi-processes/issues/105#issuecomment-720545142

**Comment by jerstlouis** [comment]
Other relevant comments from Christophe:

> I really believe that the input/ouptut definitions are way too rooted in the past XML paradigm, and a few minor changes relying on OpenAPI schema would benefit to the API (https://swagger.io/specification/#schema-object).

Elsewhere in OGC API specifications we are adopting "snippets" of JSON Schema (e.g. Tiles, Features queryables).
Could JSON Schema be re-used here in the process description to replace / simplify some elements of the description?

My previous comments on this issue:

Two things in particular which confused our developers and that I find out of place:

- The `input` inside `inputDescription` -- couldn't its members be included directly in the `inputDescription`?
- `literalDataDomain` can be completely different things: an array, an object (just for the purpose of saying it can be any value), or a string (to point a type URI if understand correctly?). That makes it quite tricky to parse in our approach to parsing JSON, and it also seems difficult to validate or present the user with valid values with all these different ways of specifying valid values for literals.
- Also, since we are discussing in #95 merging `ComplexData` and `LiteralData`, we would also need to merge `LiteralDataType` and `ComplexDataType` to simplify things and keep consistency? (e.g. being able to specify the valid formats for an inline value)

I believe this would also address https://github.com/opengeospatial/ogcapi-processes/issues/119 from a fundamental perspective.

**Comment by jerstlouis** [comment]
After reviewing the limits and objectives of the current input data type description, in terms of:
- the data types it can and cannot describe,
- the intent to provide continuity with the WPS concepts of _Literal_, _BoundingBox_ and _Complex_ data types,
- the lack of clarity on how _Complex_ data types are described in practice in _OGC API - Processes_,
- the need for a consistent approach to describe data types across the OGC API specifications,
- the widespread adoption of JSON Schema in mainstream IT for this purpose,
- the ease with which implementers can implement such an approach based on its simplicity, comprehensiveness and existing libraries,

what I propose is replacing the `input` property, which currently states `oneof` `complexDataType` `literalDataType` or `boundingBoxDataType`, with a `schema` property referencing the [JSONSchema meta-schema](https://raw.githubusercontent.com/json-schema-org/json-schema-spec/master/meta/validation.json), or a specific subset of it.

This would also allow to describe the data types inherited from WPS very easily:

**String literal**:
_(example with enumeration)_
```
    "something" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "schema" : { "type" : "string", "enum" : [ "CanBeThis", "CanBeThat" ] }
    },
```

**Numbers**:
_(real input greater than 0 and smaller or equal to 10)_

```
    "someNumber" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "schema" : {
          "type" : "number", "format" : "double",
          "minimum" : 0, "maximum" : 10, "exclusiveMinimum" : true
        } }
    },
```

**BoundingBox**
_(as a numbers array, e.g. for 4-6 items)_

```
    "bbox" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "schema" : {
          "type" : "array",
          "minItems" : 4, "maxItems" : 6,
          "items" : { "type" : "number", "format" : "double" }
        }
    },
```

**Complex JSON Structure**:
_(e.g. an ,y,z point)_

```
    "point" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "schema" : {
          "type" : "object",
          "required" : [ "x", "y", "z" ],
          "properties" : {
             "x" : { "type" : "number"},
             "y" : { "type" : "number"},
             "z" : { "type" : "number"}
           }
        }
    },
```

**Binary format**:
_(e.g. GeoTIFF media type)_

```
    "something" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "formats" : [ { "default" : true, "mediaType" : "image/tiff; application=geotiff } ],
       "schema" : { "type" : "string", "format" : "binary" }
    },
```

**JSON Format**:
_(e.g. GeoJSON media type)_

```
    "something" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "formats" : [ { "default" : true, "mediaType" : "application/geo+json" } ],
       "schema" : { "type" : "object" }
    },
```

***JSON Format***
(e.g. GeoJSON FeatureCollection):

```
    "something" : {
      "minOccurs" : 1, "maxOccurs" : 1,
       "formats" : [ { "default" : true, "mediaType" : "application/geo+json" } ],
       "schema" : { $ref" : "https://geojson.org/schema/FeatureCollection.json" }
    },
```

**Array of strings**:
_(can this be done the WPS way?)_

```
    "sourceBands" : {
      "minOccurs" : 1, "maxOccurs" : "unlimited",
       "schema" : { "type" : "array", "items" : { "type" : "string" } }
    },
```

**Dictionary of strings**
_(can this be done the WPS way?)_

```
    "bandsPythonFunctions" : {
      "minOccurs" : 1, "maxOccurs" : "unlimited",
       "schema" : { "type" : "object", "additionalProperties" : { "type" : "string" } }
    },
```

**Comment by jerstlouis** [comment]
We clarified in the meeting today that Complex data types in the current approach can specify a schema, but only as an external reference, which the current OpenAPI definition only qualifies as a string:

https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/format.yaml#L7

If I understand correctly, this means that to describe scenarios 4 (xyz point), 8 (array of strings) and 9 (dictionary of strings) above, it is not possible to do this directly in the the input description itself.
As relatively simple as they are, those data types need to be described in an external schema.
If schema was allowed to be an object, I guess one could embed a JSON schema there.

If we stick with the current approach as opposed to this JSON Schema proposal, there are still important issues that requires fixes in my opinion:

1. The [`input`](https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/inputDescription.yaml#L5) seems one of these redundant unncessary level of object properties introduced for the purpose of describing things in OpenAPI, but actually not needed (like other things that were recently fixed) 
2. The `literalDataDomain` is a real pain to deal with, partly because it can be an object (only for the purpose of specifying `{ "anyValue" : true}` , an array (of allowed values) or a string (values reference).
2.1 Why is this `anyValue` required? Couldn't any values simply be allowed if nothing is specified?
2.2 The schema could be improved instead to have dedicated members for `allowedValues`, `allowedRanges` and `valueReference` that could replace `valueDefinition` and avoid the issue of properties being different basic types.
2.3 [nameReferenceType](https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/nameReferenceType.yaml) should specifically identify the valid values as an enumeration: `integer`, `float`, `double`, `decimal`, `bool` and `string`.

@pvretano @bpross-52n

**Comment by joanma747** [comment]
I'm in favor of adopting JSON schema to describe the "nature" of any type of input in any OGC API.

Please take a look at we we do in the TileSetMetadata in the 2DTMS standard (draft). We are using JSON schema to define the properties of the features that are part of a tiles. This is done by embedding JSON schema in a JSON document. This is not our idea; OpenAPI is also doing this.

See: 
https://github.com/opengeospatial/2D-Tile-Matrix-Set/blob/master/standard/clause_9_json_encoding.adoc (Table 2. propertiesSchema attributes and JSON Schema properties equivalences)
and:
https://github.com/opengeospatial/2D-Tile-Matrix-Set/blob/master/schemas/tms/2.0/json/tileSet.json   (propertiesSchema)

I believe using JSON schema to describe the inputs of a process move us closer to what OpenAPI 3.1 is doing 
http://spec.openapis.org/oas/v3.1.0#parameter-object

See how similar the examples of JSON schema in OpenAPI 3.1 look to the @jerstlouis examples above:
http://spec.openapis.org/oas/v3.1.0#schema-object-examples

**Comment by cportele** [comment]
I would welcome the move from an OGC-specific schema language for inputs and outputs to an existing schema language that most OGC API implementations will have to deal with one way or the other anyhow (JSON Schema).

This is also the direction Features (and Records) is using. Currently this is spread over different documents, but there is a Features API SWG charter update in the approval process to create a separate part for API building blocks that are schemas; everything except the bindings to the features/records resources could eventually move to Common. 

The idea to use JSON Schema where possible/applicable. Currently we have the following:

* Response schema (schema of the feature you get back at `/collections/{collectionId}/items/{itemId}`), to be specified in the new part once the task is approved
* Queryables schema (object schema with each property that can be used in filter expressions as a property of the object), [currently in Features Part 3](http://docs.opengeospatial.org/DRAFTS/19-079.html#filter-queryables)
* Create schema (schema of a new feature sent via POST to `/collections/{collectionId}/items`), [currently in Features Part 4](http://docs.opengeospatial.org/DRAFTS/20-002.html#req_features_representation-geojson-create)
* Replace schema (schema of an updated feature sent via PUT to `/collections/{collectionId}/items/{itemId}`), [currently in Features Part 4](http://docs.opengeospatial.org/DRAFTS/20-002.html#req_features_representation-geojson-replace)
* Update schema (schema of a partial feature update sent via PATCH to `/collections/{collectionId}/items/{itemId}`), [currently in Features Part 4](http://docs.opengeospatial.org/DRAFTS/20-002.html#req_features_representation-geojson-update)
* Sortables schema (object schema with each property that can be used for sorting result as a property of the object), [currently in Records Part 1](http://docs.opengeospatial.org/DRAFTS/20-004.html#sorting-parameter-sortby)

**Comment by bpross-52n** [comment]
This should be fixed by #141
